File: Exercícios/Imovel.java
import java.time.LocalDate;

/**
 * Escreva uma descrição da classe Imovel aqui.
 * 
 * @author (João Gabriel Lofiego) 
 * @version (11/06/2023)
 */
public abstract class Imovel implements Agendavel{
    // variáveis de instância - substitua o exemplo abaixo pelo seu próprio
    private long iptu;
    private String type;
    private String util;
    private Endereco address;
    private Agenda agenda;

    /**
     * Construtor para objetos da classe Imovel
     */
    public Imovel(long iptu, String street, int number, long cep, String state, String city, String type, String util){
        // inicializa variáveis de instância
        this.iptu = iptu;
        this.type = type;
        this.util = util;
        this.address = new Endereco(street, number, cep, state, city);
        this.agenda = new Agenda();
    }

    //Construtor caso não tenha estado e cidade
    public Imovel(long iptu, String street, int number, long cep, String type, String util){
        this(iptu, street, number, cep, "BA", "Salvador", type, util);
    }

    //Métodos get

    public long getIptu(){
        return this.iptu;
    }

    public String getType(){
        return this.type;
    }

    public String getUtil(){
        return this.util;
    }

    public Endereco getAddress(){
        return this.address;
    }

    public Agenda getAgenda(){
        return this.agenda;
    }

    // Métodos sets

    public void setIptu(long iptu){
        this.iptu = iptu;
    }

    public void setType(String type){
        this.type = type;
    }

    public void setUtil(String util){
        this.util = util;
    }

    public void updateAdress(String street, int number, long cep, String state, String city){
        this.address.updateAdress(street, number, cep, state, city);
    }

    public void updateAdress(String street, int number, long cep){
        this.address.updateAdress(street, number, cep);
    }

    public abstract float calculaAluguelBase();

    @Override
    public boolean estaDisponivel(LocalDate inicio, LocalDate fim){
        for(LocalDate date : this.agenda.getDatas()){
            if(date.isAfter(inicio) && date.isBefore(fim)){
                return false;
            }
        }
        return true;
    }

    @Override
    public float valorTotal(LocalDate inicio, LocalDate fim){
        long qtdDias = fim.toEpochDay() - inicio.toEpochDay();
        
        return qtdDias * this.calculaAluguelBase();
    }

    @Override
    public float valorTotal(LocalDate data){
        return this.calculaAluguelBase();
    }

    public String toString(){
        String str = "\n===========================";
            str += "\nIPTU: " + this.iptu;
            str += "\nTipo: " + this.type;
            str += "\nUtilidade: " + this.util;
            str += "\n" + this.address.toString();
        return str;
    }
}

File: Trabalho final/Imovel.py
from Endereco import Endereco

class Imovel:
    
    __iptu : int = None
    __tipo: str = None
    __util: str = None
    __endereco: Endereco = None
    
    def __init__(self, iptu: int, tipo: str, util: str, rua: str, numero: int, cep: int, cidade: str = 'Salvador', estado: str = 'BA'):
        self.__iptu = iptu
        self.__tipo = tipo
        self.__util = util
        self.__endereco = Endereco(rua, numero, cidade, estado, cep)
        
    def __str__(self) -> str:
        string = ''
        string += '=' * 30
        string += f'\nIPTU: {self.__iptu}'
        string += f'\nTipo: {self.__tipo}'
        string += f'\nUtil: {self.__util}'
        string +=  '\n' + str(self.__endereco)
        string += '\n' + '=' * 30
        return string
        
    def getIptu(self) -> int:
        return  self.__iptu
    
    def getTipo(self) -> str:
        return  self.__tipo
    
    def getUtil(self) -> str:
        return  self.__util
    
    def getEndereco(self) -> Endereco:
        return self.__endereco
    
    def setIptu(self, iptu: int) -> None:
        self.__iptu = iptu
        
    def setTipo(self, tipo: int) -> None:
        self.__tipo = tipo
        
    def setUtil(self, util: int) -> None:
        self.__util = util
        
    def setEnderecoFull(self, rua: str, numero: int, cep: int, cidade: str, estado: str):
        self.__endereco.setRua(rua)
        self.__endereco.setNumero(numero)
        self.__endereco.setCep(cep)
        self.__endereco.setEstado(estado)
        self.__endereco.setCidade(cidade)
        
    def setEndereco(self, rua: str, numero: int, cep: int):
        self.__endereco.setRua(rua)
        self.__endereco.setNumero(numero)
        self.__endereco.setCep(cep)

#Teste Imovel    
if __name__ == '__main__':
    im = Imovel(342, 'casa', 'de praia', 'alguma ai', 12, 45032010)
    
    print(im)
    
    im.setEndereco('talvez', 90, 490940)
    
    print(im)